feat: Add support for exporting trial by label

bradbeam opened the pull request:
This also changes the UX slightly in that youll need to specify -l or -n
for the label or name selector.

will rebase off #110 when that's ready

```
redskyctl export -f $recipes/postgres/experiment.yaml -l brad=true
...
spec:
...
        resources:
          limits:
            cpu: 810m
            memory: 1056Mi
          requests:
            cpu: 810m
            memory: 1056Mi
...
```

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

bradbeam committed:
refactor: Make authorizationIdentity the default way to discover a client id

This makes the authorizationIdentity function previously defined in redskyctl the default
for getting a client identity. If a ClientIdentify function is specified, we will honor
that instead.

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

bradbeam committed:
refactor: Expose some methods from experiments so they can be used elsewhere

This is a temporary fix until the code gets migrated to a separate package to better
facilitate reuse.

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

bradbeam committed:
refactor: Create redskyctl/util package

Moving some more generic/higher level functions into the util package for reusability.

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

bradbeam committed:
refactor: Allow ReadExperiments to flex on input filetype

This changes ReadExperiments to use yaml.NewYamlOrJsonDecoder so that we can
remove the filename suffix codepath. This allows us to use a tempfile for testing
( which wouldnt contain a suffix ) as well as removes some of the fragility of
file type/media type detection.

As part of this change, we also change the function signature to a singleton. We
dont support multiple experiments defined in a single file, so we shouldnt return
an experiment list.

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

bradbeam committed:
feat: Add support for exporting trials

this is ugly/wip, but this gets us a workflow such that
- we can fetch a given trial from the api
- we read in an experiment
- we get the patch(es) from the experiment
- we fetch the object targeted by the patch targetRef
- we apply the patches using the trial data parameters to the object targeted

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

bradbeam committed:
refactor: Export name struct and rename to Identifer

We are going to be reusing the name struct for the export lookup,
so we need to be able to explicitly reference this type to pass
it around as arguments.
The rename was done to reduce the duplicaion of Name.Name and help
put a bit more meaning/context with the struct.

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

bradbeam committed:
refactor: Change SetExperiementsAPI to NewExperiementsAPI

This changes the function signature to reduce unnecessary arguments and better
align with what it is actually doing.

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

bradbeam committed:
refactor: Refactor export to support testing

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

bradbeam committed:
tests: Add additional tests for export

This should round out the test coverage for export.

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

bradbeam committed:
fix: Add IOStreams to export

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

bradbeam committed:
fix: Expose kubectl output on error

Previously we would have just failed with `Error: exit status 1`, now
we provide context ( by reporting the actual kubectl error )

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

bradbeam committed:
chore: Mark export.filename flag as required

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

bradbeam committed:
refactor: Allow kustomize to be more composable

This removes the hardcoded restriction on using embedded assets only and
exposes WithAssets and WithPatches to allow our kustomize wrapper to be
more reusable.

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

bradbeam committed:
feat: Add support for exporting trial by label

This also changes the UX slightly in that youll need to specify -l or -n
for the label or name selector.

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

jgustie commented on redskyctl/internal/commands/export/export.go:
```diff
@@ -46,8 +46,10 @@ func NewCommand(cfg config.Config) *cobra.Command {
+	cmd.Flags().StringVarP(&ec.filename, "experiment", "f", "", "path to experiment file")
+	cmd.Flags().StringVarP(&ec.trialName, "trialname", "t", "", "name of trial (ex. postgres-123)")
+	cmd.Flags().StringToStringVarP(&ec.label, "label", "l", nil, "label selector for a trial in k=v format (ex. best=true)")
+	cmd.MarkFlagRequired("experiment")
```
I'm jumping ahead of #110.

The flag names should be `--filename/-f`, `--selector/-l` for consistency. Also, object names are not typically given option names, but rather accepted as arguments (all of the "parse names" nonsense to ensure that level of consistency with expectations for users of `kubectl`).

bradbeam committed:
Revert "refactor: Make authorizationIdentity the default way to discover a client id"

This reverts commit f768f4ae5c8f2966ef51ee574c16be8ec8fea255.

bradbeam committed:
chore: fixups from review

- change exportcommand to options
- add in missing io.Reader for ReadExperiments
- change to fmt.Fprintln for output

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>
